src/internal/pfsdb/commits.go

```diff
@@ -803,6 +799,100 @@ func ListCommit(ctx context.Context, db *pachsql.DB, filter CommitListFilter, re
 	return iter, nil
 }
 
+func listCommitTx(ctx context.Context, tx *pachsql.Tx, filter CommitListFilter, reverse, revision bool) (*commitIteratorTx, error) {
+	var iter *commitIteratorTx
+	limit := 100
+	page, err := listCommitPage(ctx, tx, limit, 0, filter, reverse, revision)
+	if err != nil {
+		return nil, errors.Wrap(err, "list commit tx")
+	}
+	iter = &commitIteratorTx{
+		currRevision: -1,
+		revision:     revision,
+		reverse:      reverse,
+		commits:      page,
+		limit:        limit,
+		filter:       filter,
+		tx:           tx,
+	}
+	return iter, nil
+}
+
+func (iter *commitIteratorTx) Next(ctx context.Context, dst *CommitWithID) error {
+	if dst == nil {
+		return errors.Wrap(fmt.Errorf("commit is nil"), "get next commit")
+	}
+	var err error
+	if iter.index >= len(iter.commits) {
+		iter.index = 0
+		iter.offset += iter.limit
+		iter.commits, err = listCommitPage(ctx, iter.tx, iter.limit, iter.offset, iter.filter, iter.reverse, iter.revision)
+		if err != nil {
+			return errors.Wrap(err, "list commit page")
+		}
+		if len(iter.commits) == 0 {
+			return stream.EOS()
+		}
+	}
+	if iter.revision && iter.commits[iter.index].CreatedAt.After(iter.lastTimestamp) {
+		iter.currRevision++
+		iter.lastTimestamp = iter.commits[iter.index].CreatedAt
+	}
+	row := iter.commits[iter.index]
+	commit, err := getCommitInfoFromCommitRow(ctx, iter.tx, &row)
+	if err != nil {
+		return errors.Wrap(err, "getting commitInfo from commit row")
+	}
+	*dst = CommitWithID{
+		CommitInfo: commit,
+		ID:         row.ID,
+		Revision:   iter.currRevision,
+	}
+	iter.index++
+	return nil
+}
+
+func UpdateCommitTxByFilter(ctx context.Context, tx *pachsql.Tx, filter CommitListFilter, reverse, revision bool, cb func(commitWithID CommitWithID) error) error {
+	return errors.Wrap(listCommitTxByFilter(ctx, tx, filter, reverse, revision, func(commitWithID CommitWithID) error {
+		if err := cb(commitWithID); err != nil {
+			return err
+		}
+		return UpdateCommit(ctx, tx, commitWithID.ID, commitWithID.CommitInfo)
+	}), "list commits tx by filter")
+}
+
+func ListCommitTxByFilter(ctx context.Context, tx *pachsql.Tx, filter CommitListFilter, reverse, revision bool) ([]*pfs.CommitInfo, error) {
+	var commits []*pfs.CommitInfo
+	if err := listCommitTxByFilter(ctx, tx, filter, reverse, revision, func(commitWithID CommitWithID) error {
+		commits = append(commits, commitWithID.CommitInfo)
+		return nil
+	}); err != nil {
+		return nil, errors.Wrap(err, "list commits tx by filter")
+	}
+	return commits, nil
+}
+
+func listCommitTxByFilter(ctx context.Context, tx *pachsql.Tx, filter CommitListFilter, reverse, revision bool, cb func(commitWithID CommitWithID) error) error {
```

Comment:
It looks like this is being used correctly above.  But this callback is a *huge* footgun.  The following code would be incorrect.

```go
var myResults []*CommitWithID
dbutil.WithTx(ctx, db, func(tx *sqlx.Tx) error {
  return listCommitTxByFilter(ctx, tx, filter, reverse, revision, func(cwid *CommitWithID) error {
    myResults = append(myResults, cwid).
  })
})
```

Because `dbutil.WithTx` can restart the transaction, so it's possible to have double results.

Reply:
how would you advise us to structure this code to help avoid this footgun

Reply:
Returning a slice might be one way.  It's private so it's not a huge concern.  I think the biggest risk is someone preferring the callback stylistically and exposing it through the public API, without realizing the implications.   This isn't a blocker.